# Youtube

In [10]:
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError
from oauth2client.tools import argparser
import requests
import pandas as pd

API_KEY = ''
YOUTUBE_API_SERVICE_NAME = 'youtube'
YOUTUBE_API_SERVICE_VERSION = 'v3'

youtube = build(YOUTUBE_API_SERVICE_NAME, YOUTUBE_API_SERVICE_VERSION, developerKey = API_KEY)

In [3]:
search_something = youtube.search().list(
    q = 'Netflix',
    part = 'snippet',
    order = 'relevance',
    maxResults = 1,
).execute()

In [4]:
search_something

{'kind': 'youtube#searchListResponse',
 'etag': '6YeMc4q533izzzdGffRCgrOoZpo',
 'nextPageToken': 'CAEQAA',
 'regionCode': 'KR',
 'pageInfo': {'totalResults': 1000000, 'resultsPerPage': 1},
 'items': [{'kind': 'youtube#searchResult',
   'etag': 'ZNIzk2g4RkyBLIJ9ozEqwG0SuZM',
   'id': {'kind': 'youtube#channel', 'channelId': 'UCWOA1ZGywLbqmigxE4Qlvuw'},
   'snippet': {'publishedAt': '2012-07-17T20:45:36Z',
    'channelId': 'UCWOA1ZGywLbqmigxE4Qlvuw',
    'title': 'Netflix',
    'description': "Dearest gentle reader... Netflix is one of the world's leading entertainment services with 270 million paid memberships in over 190 ...",
    'thumbnails': {'default': {'url': 'https://yt3.ggpht.com/c5D1xS3wEG2XDLhBaW_4VCJ081qcT5rxo7X3DO-0zabsBntw_D58KwiTXThWZaizcYNmpSMhRQ=s88-c-k-c0xffffffff-no-rj-mo'},
     'medium': {'url': 'https://yt3.ggpht.com/c5D1xS3wEG2XDLhBaW_4VCJ081qcT5rxo7X3DO-0zabsBntw_D58KwiTXThWZaizcYNmpSMhRQ=s240-c-k-c0xffffffff-no-rj-mo'},
     'high': {'url': 'https://yt3.ggpht.com

In [5]:
search_something['items'][0]

{'kind': 'youtube#searchResult',
 'etag': 'ZNIzk2g4RkyBLIJ9ozEqwG0SuZM',
 'id': {'kind': 'youtube#channel', 'channelId': 'UCWOA1ZGywLbqmigxE4Qlvuw'},
 'snippet': {'publishedAt': '2012-07-17T20:45:36Z',
  'channelId': 'UCWOA1ZGywLbqmigxE4Qlvuw',
  'title': 'Netflix',
  'description': "Dearest gentle reader... Netflix is one of the world's leading entertainment services with 270 million paid memberships in over 190 ...",
  'thumbnails': {'default': {'url': 'https://yt3.ggpht.com/c5D1xS3wEG2XDLhBaW_4VCJ081qcT5rxo7X3DO-0zabsBntw_D58KwiTXThWZaizcYNmpSMhRQ=s88-c-k-c0xffffffff-no-rj-mo'},
   'medium': {'url': 'https://yt3.ggpht.com/c5D1xS3wEG2XDLhBaW_4VCJ081qcT5rxo7X3DO-0zabsBntw_D58KwiTXThWZaizcYNmpSMhRQ=s240-c-k-c0xffffffff-no-rj-mo'},
   'high': {'url': 'https://yt3.ggpht.com/c5D1xS3wEG2XDLhBaW_4VCJ081qcT5rxo7X3DO-0zabsBntw_D58KwiTXThWZaizcYNmpSMhRQ=s800-c-k-c0xffffffff-no-rj-mo'}},
  'channelTitle': 'Netflix',
  'liveBroadcastContent': 'none',
  'publishTime': '2012-07-17T20:45:36Z'}}

In [6]:
# 채널 아이디 값
channelId = 'UC920m3pMPH45qztdhppZhwA'

In [7]:
id = 'UCWOA1ZGywLbqmigxE4Qlvuw'

request = youtube.channels().list(part="snippet,contentDetails,statistics", id=id)
response = request.execute()                                    

channel_overview = {
    'title' : response['items'][0]['snippet']['title'],
    'description' : response['items'][0]['snippet']['description'],
    'publishedAt' : response['items'][0]['snippet']['publishedAt'],
    'viewCount' : response['items'][0]['statistics']['viewCount'],
    'subscriberCount' : response['items'][0]['statistics']['subscriberCount'],
    'videoCount' : response['items'][0]['statistics']['videoCount'],
    'uploads' : response['items'][0]['contentDetails']['relatedPlaylists']['uploads']
    }

df_channel_overview = pd.DataFrame([channel_overview])
df_channel_overview

title                                        description  \
0  Netflix  Dearest gentle reader...\n\n\n\nNetflix is one...   

            publishedAt   viewCount subscriberCount videoCount  \
0  2012-07-17T20:45:36Z  8522942621        28400000       7337   

                    uploads  
0  UUWOA1ZGywLbqmigxE4Qlvuw

nextPageToken
- nextPageToken은 Google API에서 페이지네이션(결과를 여러 페이지로 분할하여 반환)을 구현하는 데 사용되는 토큰입니다. 이 토큰은 이전 페이지에서 반환된 결과의 다음 페이지를 요청할 때 사용됩니다.

In [8]:
# Get all the video IDs from the channel and put them in a list
playlistId = df_channel_overview['uploads'].iloc[0]
video_ids = []

request = youtube.playlistItems().list(part="snippet,contentDetails", playlistId=playlistId, maxResults = 50)
response = request.execute()
nextPageToken = response.get('nextPageToken')

for item in response['items']:
    video_ids.append(item['contentDetails']['videoId'])

while nextPageToken is not None:
    request = youtube.playlistItems().list(part="snippet,contentDetails", playlistId=playlistId, maxResults = 50, pageToken = nextPageToken)
    response = request.execute()
    nextPageToken = response.get('nextPageToken')
    for item in response['items']:
        video_ids.append(item['contentDetails']['videoId'])

In [9]:
# Put video details in data frame
videos = []

for i in range(0, len(video_ids), 50):
    request = youtube.videos().list(part="snippet,contentDetails,statistics", id=video_ids[i:i+50])
    response = request.execute()

    for item in response['items']:
        video = {
            'channelTitle' : df_channel_overview['title'].iloc[0],
            'videoId' : item['id'],
            'categoryId' : item['snippet']['categoryId'],
            'publishedAt' : item['snippet']['publishedAt'],
            'title' : item['snippet']['title'],
            'description' : item['snippet']['description'],
            'tags' : item['snippet'].get('tags','no_tags'),
            'duration' : item['contentDetails']['duration'],
            'viewCount' : item['statistics'].get('viewCount',0),
            'likeCount' : item['statistics'].get('likeCount', 0),
            'commentCount' : item['statistics'].get('commentCount',0)
        }
        videos.append(video)
    
df_videos = pd.DataFrame(videos)
len(df_videos)

7337

In [10]:
# https://sddev.tistory.com/88
import isodate
# Convert duration column to seconds
df_videos['duration_sec'] = df_videos['duration'].apply(lambda x: isodate.parse_duration(x).total_seconds())

# Convert specific columns to numeric type
numeric_columns = ['viewCount', 'likeCount', 'commentCount', 'duration_sec']
df_videos[numeric_columns] = df_videos[numeric_columns].apply(pd.to_numeric, errors='coerce')

# Convert column to Datetime to access the year
df_videos['publishedAt'] = pd.to_datetime(df_videos['publishedAt'])
df_videos['year'] = df_videos['publishedAt'].dt.year

In [11]:
df_videos.head()

channelTitle      videoId categoryId               publishedAt  \
0      Netflix  EnuUS9OjV7c         24 2024-06-14 13:00:00+00:00   
1      Netflix  nAYKaslCXPc         24 2024-06-14 02:00:00+00:00   
2      Netflix  D5JXgPrdfsw         24 2024-06-13 15:00:00+00:00   
3      Netflix  FQoaul1bSF4         24 2024-06-13 14:00:00+00:00   
4      Netflix  iFKTVEoApwo         24 2024-06-13 13:30:00+00:00   

                                               title  \
0  Ultraman Vs. Gigantron | Ultraman: Rising | Ne...   
1    Godzilla Minus One | Official Trailer | Netflix   
2  The Accidental Twins | Official Trailer | Netflix   
3              Receiver | Official Trailer | Netflix   
4  Perfect Match S2 | Jess Grills Harry | Sneak P...   

                                         description  \
0  Not only is Ultraman back… so is Gigantron! Ta...   
1  In postwar Japan, a new terror rises; Godzilla...   
2  Two sets of identical twins switched at birth ...   
3  SEE THE GAME FROM THE OTHER SIDE OF THE PASS. ...   
4  Netflix reality show alumni look for lasting l...   

                                                tags duration  viewCount  \
0  [Bandai, Gigantron, Tsuburaya, Ultraman, actio...  PT4M17S       6914   
1  [Godzilla, GodzillaMinusOne, Netflix, Takashi ...  PT1M56S     411707   
2  [Captivating, Compassionate, Inspiring, Inters...   PT2M7S      20441   
3  [49ers, AFC, Behind The Scenes, Best Plays, Cl...  PT2M46S     147692   
4  [Addictive, Breakups, Competition, Dating, Fin...  PT1M48S      16824   

   likeCount  commentCount  duration_sec  year  
0        555            37         257.0  2024  
1      13885          1295         116.0  2024  
2        337            23         127.0  2024  
3       4270           253         166.0  2024  
4        191            33         108.0  2024

In [12]:
df_videos.columns = ['채널명','영상ID', '카테고리ID', '게시일' ,"제목", '설명', '출연자_정보', '재생시간'
                     ,'조회수', '좋아요수', '댓글수',  '재생시간(초)', '년도'] 

In [13]:
df_videos['년도'].value_counts()

년도
2022    1000
2021     965
2020     943
2019     927
2018     847
2023     841
2017     580
2024     451
2016     363
2015     171
2013     127
2014     109
2012      13
Name: count, dtype: int64

In [15]:
# publishedAt 열을 datetime 형식으로 변환
df_videos['게시일'] = pd.to_datetime(df_videos['게시일']).dt.tz_localize(None)

df_videos.to_csv('../data/file/df_videos.csv', sep=",", encoding="utf-8")

# df_videos.to_excel('../data/file/df_videos.xlsx',index=False)

---

In [12]:
df_videos = pd.read_csv('../data/file/df_videos.csv', index_col = 0)

In [13]:
trailer = df_videos[df_videos['제목'].str.contains('Trailer|trailer', case=False)]

In [14]:
df_videos[df_videos['제목'].str.contains('JIVA')]

채널명         영상ID  카테고리ID                  게시일  \
2861  Netflix  zGYOt0ZmifQ      24  2021-06-07 08:00:00   
2893  Netflix  Y9LaKjImxBA      24  2021-05-24 08:00:00   

                                      제목  \
2861  JIVA! | Official Trailer | Netflix   
2893   JIVA! | Official Teaser | Netflix   

                                                     설명  \
2861  A talented street dancer struggles while juggl...   
2893  "This is our chance to shine." As'hambeni. HAY...   

                                                 출연자_정보     재생시간     조회수  \
2861  ['Netflix series', 'dance', 'jiva', 'netflix 2...  PT2M11S  316768   
2893  ['Netflix series', 'dance', 'jiva', 'netflix 2...    PT59S   70957   

      좋아요수  댓글수  재생시간(초)    년도  
2861  6552  335    131.0  2021  
2893  2208  195     59.0  2021

---

In [15]:
# 엑셀 파일을 읽어옵니다.
df = pd.read_excel('../data/file/imdb_final_4.xlsx', index_col=0)
df.head()

korean_title      english_title  year  total_season_num  runtime  \
0              오징어 게임         Squid Game  2021                 1       55   
1            12인의 심판자         The Twelve  2019                 2       54   
2              희생자 게임  The Victims' Game  2020                 1       60   
3  스위트 투스: 사슴뿔을 가진 소년        Sweet Tooth  2021                 2       52   
4                스위트홈         Sweet Home  2020                 2       60   

                       genre age_rating production_country  total_count  \
0               액션, 드라마, 스릴러         19               대한민국       549000   
1                   드라마, 스릴러         19                벨기에         6500   
2               드라마, 범죄, 스릴러         19                 대만         1900   
3           SF, 드라마, 판타지, 액션         15                 미국        68000   
4  스릴러, 드라마, 판타지, 공포, SF, 범죄         19               대한민국        33000   

   total_rate  ...  season_13  season_14  season_15  season_16 season_17  \
0         8.0  ...        NaN        NaN        NaN        NaN       NaN   
1         7.4  ...        NaN        NaN        NaN        NaN       NaN   
2         7.3  ...        NaN        NaN        NaN        NaN       NaN   
3         7.7  ...        NaN        NaN        NaN        NaN       NaN   
4         7.3  ...        NaN        NaN        NaN        NaN       NaN   

                                            director  \
0                                ['Hwang Dong-hyuk']   
1                   ['Kaat Beels', 'Wouter Bouvijn']   
2                ['David Chuang', 'Kuan-Chung Chen']   
3  ['Alexis Ostrander', 'Carol Banker', 'Ciarán F...   
4  ['Jang Young-woo', 'Lee Eung-bok', 'Park So-hy...   

                                              writer  \
0                                ['Hwang Dong-hyuk']   
1  ['Bert Van Dael', 'Nele Meirhaeghe', 'Roel Mon...   
2  ['Joyce Liu', 'Jui-Liang Hsu', 'Shih-Keng Chie...   
3  ['Beth Schwartz', 'Bo Yeon Kim', 'Carly Woodwo...   
4  ['Hong So-ri', 'Hwang Young-Chan', 'Kim Hyung-...   

                                               actor  \
0  ['Lee Jung-jae', 'Park Hae-soo', 'Yasushi Iwak...   
1  ['Luc De Ruelle', 'Maaike Neuville', 'Tom Verm...   
2  ['Wei-Ning Hsu', 'Hsiao-chuan Chang', 'Shih-Si...   
3  ['Nonso Anozie', 'Christian Convery', 'Stefani...   
4  ['Song Kang', 'Lee Jin-wook', 'Lee Si-young', ...   

                           imdb_title                                imdb_url  
0         Squid Game (TV Series 2021)  https://www.imdb.com/title/tt10919420/  
1    The Twelve (TV Series 2019–2023)   https://www.imdb.com/title/tt7605396/  
2  The Victims' Game (TV Series 2020)  https://www.imdb.com/title/tt12079212/  
3        Sweet Tooth (TV Series 2021)  https://www.imdb.com/title/tt12809988/  
4         Sweet Home (TV Series 2020)  https://www.imdb.com/title/tt11612120/  

[5 rows x 38 columns]

In [16]:
trailer.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2947 entries, 1 to 7324
Data columns (total 13 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   채널명      2947 non-null   object 
 1   영상ID     2947 non-null   object 
 2   카테고리ID   2947 non-null   int64  
 3   게시일      2947 non-null   object 
 4   제목       2947 non-null   object 
 5   설명       2947 non-null   object 
 6   출연자_정보   2947 non-null   object 
 7   재생시간     2947 non-null   object 
 8   조회수      2947 non-null   int64  
 9   좋아요수     2947 non-null   int64  
 10  댓글수      2947 non-null   int64  
 11  재생시간(초)  2947 non-null   float64
 12  년도       2947 non-null   int64  
dtypes: float64(1), int64(5), object(7)
memory usage: 322.3+ KB


In [17]:
import re  # re 모듈을 임포트

# youtube_df 초기화
youtube_df = pd.DataFrame()

# for 문을 사용하여 필터링
for title in df['english_title']:
    try:
        # 정규 표현식 이스케이프를 사용하여 title 처리
        escaped_title = re.escape(title)
        filtered_df = trailer[trailer['제목'].str.contains(escaped_title, case=False)]
        
        # 필터링된 결과가 없는 경우에도 해당 영화 제목 출력
        if filtered_df.empty:
            print(f"No data found for '{title}' Trailer")
        else:
            youtube_df = pd.concat([youtube_df, filtered_df], ignore_index=True)
            print(f"Filtered data for '{title}' Trailer:")
            print(filtered_df)
    except Exception as e:
        print(f"An error occurred while filtering for title '{title}': {e}")

# 필터링된 결과 확인
print("\nFinal filtered dataframe:")
print(youtube_df)

Filtered data for 'Squid Game' Trailer:
          채널명         영상ID  카테고리ID                  게시일  \
515   Netflix  W-Cc6XaneLs      24  2023-12-01 15:00:04   
516   Netflix  OzJ1aS317mg      24  2023-12-01 15:00:00   
653   Netflix  O61C8zc8Znk      24  2023-10-23 15:00:01   
2588  Netflix  oqxAJKy0ii4      24  2021-09-02 00:00:02   

                                                     제목  \
515   Squid Game: The Challenge | Who Will Win the 4...   
516   Squid Game: The Challenge | Who Will Win the 4...   
653   Squid Game: The Challenge | Official Trailer |...   
2588            Squid Game | Official Trailer | Netflix   

                                                     설명  \
515   The game that started with 456 players competi...   
516   The game that started with 456 players competi...   
653   Make friends. Make enemies. Make millions. Squ...   
2588  A Netflix Series | Squid Game \nSurvive or die...   

                                                 출연자_정보     재생시간       조

In [18]:
# youtube_df 및 no_data_df 초기화
youtube_df = pd.DataFrame()
no_data_df = pd.DataFrame(columns=['Title'])

# for 문을 사용하여 필터링
for title in df['english_title']:
    try:
        # 정규 표현식 이스케이프를 사용하여 title 처리
        escaped_title = re.escape(title)
        filtered_df = trailer[trailer['제목'].str.contains(escaped_title, case=False)]
        
        # 필터링된 결과가 없는 경우와 있는 경우를 분리하여 처리
        if filtered_df.empty:
            no_data_df = pd.concat([no_data_df, pd.DataFrame({'Title': [title]})], ignore_index=True)
        else:
            youtube_df = pd.concat([youtube_df, filtered_df], ignore_index=True)
    except Exception as e:
        print(f"An error occurred while filtering for title '{title}': {e}")

In [19]:
# 필터링된 결과가 없는 경우와 있는 경우의 데이터프레임 출력
print("Filtered dataframe with data:")
print(youtube_df)

Filtered dataframe with data:
          채널명         영상ID  카테고리ID                  게시일  \
0     Netflix  W-Cc6XaneLs      24  2023-12-01 15:00:04   
1     Netflix  OzJ1aS317mg      24  2023-12-01 15:00:00   
2     Netflix  O61C8zc8Znk      24  2023-10-23 15:00:01   
3     Netflix  oqxAJKy0ii4      24  2021-09-02 00:00:02   
4     Netflix  Os4R9IEBOA8      24  2024-05-21 16:00:03   
...       ...          ...     ...                  ...   
1266  Netflix  5yxjRgwYymg      24  2019-05-17 03:30:01   
1267  Netflix  W-Cc6XaneLs      24  2023-12-01 15:00:04   
1268  Netflix  OzJ1aS317mg      24  2023-12-01 15:00:00   
1269  Netflix  iDvPvqImb-4      24  2021-09-30 09:00:00   
1270  Netflix  KmtAZKXb6kU      24  2020-08-12 18:00:00   

                                                     제목  \
0     Squid Game: The Challenge | Who Will Win the 4...   
1     Squid Game: The Challenge | Who Will Win the 4...   
2     Squid Game: The Challenge | Official Trailer |...   
3               Squid Gam

In [20]:
print("\nTitles with no data found:")
print(no_data_df)


Titles with no data found:
                                 Title
0                           The Twelve
1                    The Victims' Game
2                      The Good Doctor
3                     The Walking Dead
4                            The Flash
..                                 ...
501   Close Your Eyes Before It's Dark
502                 Bhaag Beanie Bhaag
503                  Million Yen Women
504                            Hasmukh
505  Elite Short Stories: Nadia Guzmán

[506 rows x 1 columns]


- 찾으려는 데이터 506개가(900개 중) Youtube Netflix 공식계정에 없음
- 다른방법을 강구

---

In [2]:
import pandas as pd
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError
import time

def get_search_results_count(query):
    youtube = build('youtube', 'v3', developerKey=API_KEY)
    total_results = 0
    next_page_token = None
    max_retries = 5  # 최대 재시도 횟수 설정

    while True:
        try:
            response = youtube.search().list(
                part='id',
                q=query,
                type='video',
                maxResults=50,
                pageToken=next_page_token
            ).execute()

            total_results += len(response['items'])
            next_page_token = response.get('nextPageToken')

            if not next_page_token:
                break

        except HttpError as e:
            if e.resp.status == 403:
                for attempt in range(max_retries):
                    print(f"Quota exceeded. Retrying in 5 minutes... (Attempt {attempt + 1}/{max_retries})")
                    time.sleep(300)  # 5분 대기
                    try:
                        response = youtube.search().list(
                            part='id',
                            q=query,
                            type='video',
                            maxResults=50,
                            pageToken=next_page_token
                        ).execute()
                        total_results += len(response['items'])
                        next_page_token = response.get('nextPageToken')
                        if not next_page_token:
                            break
                    except HttpError as e:
                        if e.resp.status == 403:
                            continue
                        else:
                            raise e
                else:
                    print("Maximum retries reached. Skipping this query.")
                    return total_results
            else:
                raise e

    return total_results

In [7]:
def get_video_details(query):
    # YouTube API 클라이언트 생성
    youtube = build('youtube', 'v3', developerKey=API_KEY)
    
    # 초기 설정
    next_page_token = None
    video_details = []

    try:
        while True:
            # 검색어에 해당하는 동영상 검색
            response = youtube.search().list(
                part='snippet',
                q=query,
                type='video',
                maxResults=50,  # 한 번에 가져올 최대 결과 수
                pageToken=next_page_token  # 페이지 토큰 설정
            ).execute()

            # 검색 결과의 동영상 정보 수집
            for item in response['items']:
                video_id = item['id']['videoId']
                title = item['snippet']['title']
                published_at = item['snippet']['publishedAt']
                video_details.append((video_id, title, published_at))

            # 다음 페이지 토큰이 있는지 확인
            next_page_token = response.get('nextPageToken')

            # 다음 페이지가 없으면 종료
            if not next_page_token:
                break

    except Exception as e:
        print(f"An error occurred: {e}")

    return video_details

# 검색어를 여기에 입력하세요
search_query = 'Jiva!'
videos = get_video_details(search_query)

# 결과 출력
for video_id, title, published_at in videos:
    print(f'Title: {title}, Published At: {published_at}, Video ID: {video_id}')

An error occurred: <HttpError 403 when requesting https://youtube.googleapis.com/youtube/v3/search?part=snippet&q=Jiva%21&type=video&maxResults=50&key=AIzaSyDPM5Z3kZi_lKfsTjxlIiMGc8UfdzCbLB8&alt=json returned "The request cannot be completed because you have exceeded your <a href="/youtube/v3/getting-started#quota">quota</a>.". Details: "[{'message': 'The request cannot be completed because you have exceeded your <a href="/youtube/v3/getting-started#quota">quota</a>.', 'domain': 'youtube.quota', 'reason': 'quotaExceeded'}]">


In [5]:
results = []

for title in df_2['english_title'][:10]:
    search_query = f"{title} season 1"
    result_count = get_search_results_count(search_query)
    results.append({'search_query': search_query, 'result_count': result_count})

df_results = pd.DataFrame(results)

Quota exceeded. Retrying in 5 minutes... (Attempt 1/5)


KeyboardInterrupt: 

In [6]:
results

[{'search_query': 'The Twelve season 1', 'result_count': 600},
 {'search_query': 'Sweet Tooth season 1', 'result_count': 597},
 {'search_query': 'Sweet Home season 1', 'result_count': 597},
 {'search_query': 'The Good Doctor season 1', 'result_count': 600},
 {'search_query': 'The Walking Dead season 1', 'result_count': 597},
 {'search_query': 'Borgen - Power & Glory season 1', 'result_count': 491},
 {'search_query': 'The Flash season 1', 'result_count': 601},
 {'search_query': 'Bridgerton season 1', 'result_count': 609}]

- 각 영화의 게시물 갯수로 흥행을 알아보려 했으나 쿼터의 하루할당량 초과와 각 드라마의 인기여부와 상관없이 게시물 갯수가 대부분 600개 정도 비슷한 수치를 보임

---

In [4]:
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError

def get_top_video(query):
    youtube = build('youtube', 'v3', developerKey=API_KEY)
    
    try:
        # 검색어에 해당하는 조회수가 높은 동영상 검색
        search_response = youtube.search().list(
            part='id,snippet',
            q=query,
            type='video',
            maxResults=1,  # 한 번에 하나의 결과만 가져오기
            order='viewCount'  # 조회수 기준으로 정렬
        ).execute()

        if not search_response['items']:
            print("No videos found for the query.")
            return None

        video_id = search_response['items'][0]['id']['videoId']
        title = search_response['items'][0]['snippet']['title']
        channel_title = search_response['items'][0]['snippet']['channelTitle']

        # 비디오의 상세 정보 가져오기
        video_response = youtube.videos().list(
            part='statistics',
            id=video_id
        ).execute()

        if not video_response['items']:
            print("No video details found.")
            return None

        view_count = int(video_response['items'][0]['statistics'].get('viewCount', 0))
        like_count = int(video_response['items'][0]['statistics'].get('likeCount', 0))

        return {
            'videoId': video_id,
            'title': title,
            'channelTitle': channel_title,
            'viewCount': view_count,
            'likeCount': like_count
        }

    except HttpError as e:
        print(f"An HTTP error {e.resp.status} occurred: {e.content}")
    except Exception as e:
        print(f"An error occurred: {e}")

    return None

# 검색어를 여기에 입력하세요
search_query = 'Jiva!'
top_video = get_top_video(search_query)

# 결과 출력
if top_video:
    print(f"Title: {top_video['title']}")
    print(f"Channel: {top_video['channelTitle']}")
    print(f"Views: {top_video['viewCount']}")
    print(f"Likes: {top_video['likeCount']}")
    print(f"Video ID: {top_video['videoId']}")
else:
    print("No popular video found.")


Title: Kee (2019) New Released Hindi Dubbed Full Movie | Jiiva, Govind Padmasoorya, Nikki Galrani
Channel: Goldmines
Views: 72052317
Likes: 710154
Video ID: hu6jla2gMXI


In [7]:
def get_top_video(query):
    youtube = build('youtube', 'v3', developerKey=API_KEY)
    
    try:
        # 검색어에 해당하는 동영상 검색
        search_response = youtube.search().list(
            part='id,snippet',
            q=query,
            type='video',
            maxResults=50,  # 최대 50개의 결과를 가져와서 필터링
            order='relevance',  # 관련성 기준으로 정렬
            relevanceLanguage='en',  # 언어 설정 (필요한 언어 코드로 변경 가능)
            regionCode='US'  # 지역 설정 (필요한 지역 코드로 변경 가능)
        ).execute()

        if not search_response['items']:
            print("No videos found for the query.")
            return None

        # 검색 결과 필터링: 제목에 검색어가 대소문자 구분 없이 포함된 경우 우선
        filtered_items = [
            item for item in search_response['items']
            if query.lower() in item['snippet']['title'].lower()
        ]

        if not filtered_items:
            filtered_items = search_response['items']

        video_id = filtered_items[0]['id']['videoId']
        title = filtered_items[0]['snippet']['title']
        channel_title = filtered_items[0]['snippet']['channelTitle']

        # 비디오의 상세 정보 가져오기
        video_response = youtube.videos().list(
            part='statistics',
            id=video_id
        ).execute()

        if not video_response['items']:
            print("No video details found.")
            return None

        view_count = int(video_response['items'][0]['statistics'].get('viewCount', 0))
        like_count = int(video_response['items'][0]['statistics'].get('likeCount', 0))

        return {
            'videoId': video_id,
            'title': title,
            'channelTitle': channel_title,
            'viewCount': view_count,
            'likeCount': like_count
        }

    except HttpError as e:
        print(f"An HTTP error {e.resp.status} occurred: {e.content}")
    except Exception as e:
        print(f"An error occurred: {e}")

    return None

# 검색어를 여기에 입력하세요
search_query = 'The Twelve trailer'
top_video = get_top_video(search_query)

# 결과 출력
if top_video:
    print(f"Title: {top_video['title']}")
    print(f"Channel: {top_video['channelTitle']}")
    print(f"Views: {top_video['viewCount']}")
    print(f"Likes: {top_video['likeCount']}")
    print(f"Video ID: {top_video['videoId']}")
else:
    print("No popular video found.")

Title: THE TWELVE TRAILER
Channel: Walter Presents UK
Views: 13453
Likes: 22
Video ID: 9Ek19Tx3Rto


In [11]:
def get_most_popular_video(query):
    youtube = build('youtube', 'v3', developerKey=API_KEY)
    
    try:
        # 검색어에 해당하는 동영상 검색
        search_response = youtube.search().list(
            part='id,snippet',
            q=query,
            type='video',
            maxResults=50,  # 최대 50개의 결과를 가져와서 필터링
            order='relevance',  # 관련성 기준으로 정렬
            relevanceLanguage='en',  # 언어 설정 (필요한 언어 코드로 변경 가능)
            regionCode='US'  # 지역 설정 (필요한 지역 코드로 변경 가능)
        ).execute()

        if not search_response['items']:
            print("No videos found for the query.")
            return None

        # 검색 결과 필터링: 제목에 검색어가 대소문자 구분 없이 포함된 경우 우선
        filtered_items = [
            item for item in search_response['items']
            if query.lower() in item['snippet']['title'].lower()
        ]

        if not filtered_items:
            filtered_items = search_response['items']

        video_ids = [item['id']['videoId'] for item in filtered_items]

        # 비디오의 상세 정보 가져오기
        video_response = youtube.videos().list(
            part='statistics,snippet',
            id=','.join(video_ids)
        ).execute()

        if not video_response['items']:
            print("No video details found.")
            return None

        # 조회수와 좋아요 수 기준으로 가장 인기 있는 비디오 찾기
        most_popular_video = None
        max_views = 0
        max_likes = 0

        for item in video_response['items']:
            view_count = int(item['statistics'].get('viewCount', 0))
            like_count = int(item['statistics'].get('likeCount', 0))
            if view_count > max_views or (view_count == max_views and like_count > max_likes):
                max_views = view_count
                max_likes = like_count
                most_popular_video = {
                    'videoId': item['id'],
                    'title': item['snippet']['title'],
                    'channelTitle': item['snippet']['channelTitle'],
                    'viewCount': view_count,
                    'likeCount': like_count
                }

        return most_popular_video

    except HttpError as e:
        print(f"An HTTP error {e.resp.status} occurred: {e.content}")
    except Exception as e:
        print(f"An error occurred: {e}")

    return None

# 검색어를 여기에 입력하세요
search_query = 'Jiva! trailer'
most_popular_video = get_most_popular_video(search_query)

# 결과 출력
if most_popular_video:
    print(f"Title: {most_popular_video['title']}")
    print(f"Channel: {most_popular_video['channelTitle']}")
    print(f"Views: {most_popular_video['viewCount']}")
    print(f"Likes: {most_popular_video['likeCount']}")
    print(f"Video ID: {most_popular_video['videoId']}")
else:
    print("No popular video found.")


Title: JIVA! Trailer (2021) Netflix Dance Series
Channel: ONE Media Coverage
Views: 36516
Likes: 649
Video ID: a5vEvin2DeI


- 특정 영화 검색어를 쿼리하여 조회수가 높은 동영상 검색(특정 채널이 아닌 모든 채널 중)하여 좋아요와 조회수를 가져오려했지만 검색의 정확도가 떨어져 해당 영화가 아닌 비슷한 단어가 들어간 가장 인기많은 게시물이 나옴